File: js/tests/serializer.test.ts
/**
 * Tests for the meta-notation serializer
 */

import { test, assert } from 'test-anywhere';
import { parse } from '../src/parser.js';
import { serialize } from '../src/serializer.js';

test('serialize plain text', () => {
  const blocks = [{ type: 'text' as const, content: 'hello world' }];
  const result = serialize(blocks);
  assert.equal(result, 'hello world');
});

test('serialize parentheses', () => {
  const blocks = [
    { type: 'paren' as const, content: [{ type: 'text' as const, content: 'hello' }] }
  ];
  const result = serialize(blocks);
  assert.equal(result, '(hello)');
});

test('serialize curly braces', () => {
  const blocks = [
    { type: 'curly' as const, content: [{ type: 'text' as const, content: 'world' }] }
  ];
  const result = serialize(blocks);
  assert.equal(result, '{world}');
});

test('serialize square brackets', () => {
  const blocks = [
    { type: 'square' as const, content: [{ type: 'text' as const, content: 'test' }] }
  ];
  const result = serialize(blocks);
  assert.equal(result, '[test]');
});

test('serialize quotes', () => {
  const blocks = [
    { type: 'singleQuote' as const, content: 'hello' },
    { type: 'text' as const, content: ' ' },
    { type: 'doubleQuote' as const, content: 'world' }
  ];
  const result = serialize(blocks);
  assert.equal(result, `'hello' "world"`);
});

test('serialize backticks', () => {
  const blocks = [
    { type: 'backtick' as const, content: 'code' }
  ];
  const result = serialize(blocks);
  assert.equal(result, '`code`');
});

test('round-trip: parse then serialize', () => {
  const original = 'hello (world) {test} [array] "string" `code`';
  const parsed = parse(original);
  const serialized = serialize(parsed);
  assert.equal(serialized, original);
});

test('round-trip: nested structures', () => {
  const original = '{a [b (c) d] e}';
  const parsed = parse(original);
  const serialized = serialize(parsed);
  assert.equal(serialized, original);
});

test('round-trip: empty delimiters', () => {
  const original = '(){}[]';
  const parsed = parse(original);
  const serialized = serialize(parsed);
  assert.equal(serialized, original);
});

test('round-trip: complex code', () => {
  const original = 'function test() { return "hello"; }';
  const parsed = parse(original);
  const serialized = serialize(parsed);
  assert.equal(serialized, original);
});

File: rust/tests/serializer_test.rs
//! Tests for the meta-notation serializer
//!
//! These tests verify that serialization produces the expected output and
//! that round-trip (parse -> serialize) preserves the original text.

use meta_notation::{parse, serialize, Block};

#[test]
fn test_serialize_plain_text() {
    let blocks = vec![Block::Text("hello world".to_string())];
    let result = serialize(&blocks);
    assert_eq!(result, "hello world");
}

#[test]
fn test_serialize_parentheses() {
    let blocks = vec![Block::Paren(vec![Block::Text("hello".to_string())])];
    let result = serialize(&blocks);
    assert_eq!(result, "(hello)");
}

#[test]
fn test_serialize_curly_braces() {
    let blocks = vec![Block::Curly(vec![Block::Text("world".to_string())])];
    let result = serialize(&blocks);
    assert_eq!(result, "{world}");
}

#[test]
fn test_serialize_square_brackets() {
    let blocks = vec![Block::Square(vec![Block::Text("test".to_string())])];
    let result = serialize(&blocks);
    assert_eq!(result, "[test]");
}

#[test]
fn test_serialize_quotes() {
    let blocks = vec![
        Block::SingleQuote("hello".to_string()),
        Block::Text(" ".to_string()),
        Block::DoubleQuote("world".to_string()),
    ];
    let result = serialize(&blocks);
    assert_eq!(result, "'hello' \"world\"");
}

#[test]
fn test_serialize_backticks() {
    let blocks = vec![Block::Backtick("code".to_string())];
    let result = serialize(&blocks);
    assert_eq!(result, "`code`");
}

#[test]
fn test_round_trip_parse_then_serialize() {
    let original = "hello (world) {test} [array] \"string\" `code`";
    let parsed = parse(original);
    let serialized = serialize(&parsed);
    assert_eq!(serialized, original);
}

#[test]
fn test_round_trip_nested_structures() {
    let original = "{a [b (c) d] e}";
    let parsed = parse(original);
    let serialized = serialize(&parsed);
    assert_eq!(serialized, original);
}

#[test]
fn test_round_trip_empty_delimiters() {
    let original = "(){}[]";
    let parsed = parse(original);
    let serialized = serialize(&parsed);
    assert_eq!(serialized, original);
}

#[test]
fn test_round_trip_complex_code() {
    let original = "function test() { return \"hello\"; }";
    let parsed = parse(original);
    let serialized = serialize(&parsed);
    assert_eq!(serialized, original);
}
